manim/cli/render/commands.py

```diff
@@ -0,0 +1,386 @@
+    write_all,
+    format,
+    save_last_frame,
+    quality,
+    resolution,
+    frame_rate,
+    opengl_renderer,
+    webgl_renderer,
+    transparent,
+    background_color,
+    progress_bar,
+    preview,
+    show_in_file_browser,
+    sound,
+    jupyter,
+):
+    """Render SCENE(S) from the input FILE.
+
+    FILE is the file path of the script.
+
+    SCENES is an optional list of scenes in the file.
+    """
+    for scene in scenes:
+        if str(scene).startswith("-"):
+            logger.warning(
+                dedent(
+                    """\
+                Manim Community has moved to Click for the CLI.
+
+                This means that options in the CLI are provided BEFORE the positional
+                arguments for your FILE and SCENE(s):
+                `manim render [OPTIONS] [FILE] [SCENES]...`
+
+                For example:
+                New way - `manim -p -ql file.py SceneName1 SceneName2 ...`
+                Old way - `manim file.py SceneName1 SceneName2 ... -p -ql`
+
+                To see the help page for the new available options, run:
+                `manim render -h`
+                """
+                )
+            )
+            sys.exit()
+    args = {
+        "ctx": ctx,
+        "file": file,
+        "scene_names": scenes,
+        "config_file": config_file,
+        "custom_folders": custom_folders,
+        "disable_caching": disable_caching,
+        "flush_cache": flush_cache,
+        "tex_template": tex_template,
+        "verbosity": verbose,
+        "output_file": output,
+        "write_to_movie": write_to_movie,
+        "media_dir": media_dir,
+        "log_dir": log_dir,
+        "log_to_file": log_to_file,
+        "from_animation_number": from_animation_number,
+        "write_all": write_all,
+        "format": format,
+        "save_last_frame": save_last_frame,
+        "quality": quality,
+        "resolution": resolution,
+        "frame_rate": frame_rate,
+        "use_opengl_renderer": opengl_renderer,
+        "webgl_renderer": webgl_renderer,
+        "transparent": transparent,
+        "background_color": background_color,
+        "progress_bar": progress_bar,
+        "preview": preview,
+        "show_in_file_browser": show_in_file_browser,
+        "sound": sound,
+        "jupyter": jupyter,
+    }
+
+    class ClickArgs:
+        def __init__(self, args):
+            for name in args:
+                setattr(self, name, args[name])
+
+        def _get_kwargs(self):
+            return list(self.__dict__.items())
+
+        def __eq__(self, other):
+            if not isinstance(other, ClickArgs):
+                return NotImplemented
+            return vars(self) == vars(other)
+
+        def __contains__(self, key):
+            return key in self.__dict__
+
+        def __repr__(self):
+            return str(self.__dict__)
+
+    click_args = ClickArgs(args)
+    if jupyter:
+        return click_args
+    config.digest_args(click_args)
```

Comment:
Not exactly, and the confusion from this stems from how `ipython_magic.py` [was using parse_args](https://github.com/ManimCommunity/manim/pull/1013/files#diff-cdb349de5c3c433c845c518de75c515d95eb3c5329537a30b603c59df1709940L74) from `main_util.py` (which is deleted in this PR).

Perhaps @behackl can chime in with a better method of doing this in Click. I only spent a few hours looking into this and there's a lot to think about here regarding plugins/jupyter and Manim. To rectify the lack of having Argparse's parse_args in `main_utils.py`, I created a `--jupyter` flag which essentially returns an Argparse.Namespace clone (click_args) early.

So in Click + jupyter-lab, manim is ran twice:

1. Run it to get an "Argparse.Namespace" lookalike
2. After parsing the ManimConfig, run the Scene directly with the Scene.render()


The config file logic in `ipython_magic.py` is done outside of the first

This logic hasn't changed, so you can still use config files as normal. See [here on L74](https://github.com/ManimCommunity/manim/pull/1013/files#diff-cdb349de5c3c433c845c518de75c515d95eb3c5329537a30b603c59df1709940L79) where `config.digest_args()` is used in both Click/Argparse

Reply:
Hmm. I agree, this workaround is a bit unfortunate. Effectively, all that the Jupyter interface needs is a convenient way to set config options, and given that the cell magic is called `%%manim`, the CLI args sort of offered themselves as a way to do so. (I still think that from a UX perspective, this makes a lot of sense: it's convenient to be able to use "the same" interface in two different situations.)

I don't know what the best approach for this would look like in Click world; but some way to simply set the config options based on the passed (CLI-like) arguments would be convenient. I think it's fine to investigate this further in a follow-up PR though.